File: item-list/src/main/java/com/gurunars/item_list/BindableViewHolder.java
package com.gurunars.item_list;

import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;

import static com.gurunars.item_list.ItemViewBinderEmpty.EMPTY_TYPE;

class BindableViewHolder<ViewType extends View, ItemType extends Item> extends RecyclerView.ViewHolder {

    private ItemViewBinder<ViewType, ItemType> itemViewBinder;

    BindableViewHolder(ViewGroup root,
                       ItemViewBinder<ViewType, ItemType> itemViewBinder) {
        super(itemViewBinder.getView(root.getContext()));
        this.itemViewBinder = itemViewBinder;
        itemView.setLayoutParams(new RecyclerView.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT));
    }

    BindableViewHolder(ViewGroup root,
                       EmptyViewBinder emptyViewBinder) {
        super(emptyViewBinder.getView(root.getContext()));
        this.itemViewBinder = null;
        itemView.setTag(EMPTY_TYPE, EMPTY_TYPE);
        itemView.setLayoutParams(new RecyclerView.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT));
    }

    void bind(ItemType item, @Nullable ItemType previousItem) {
        if (this.itemViewBinder != null) {
            itemViewBinder.bind((ViewType) itemView, item, previousItem);
        }
    }
}

File: item-list/src/main/java/com/gurunars/item_list/BindableViewHolder.kt
package com.gurunars.item_list

import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.ViewGroup
import com.gurunars.item_list.ItemViewBinderEmpty.Companion.EMPTY_TYPE
import com.gurunars.shortcuts.asRow
import com.gurunars.shortcuts.fullSize


internal class BindableViewHolder<ViewType : View, ItemType : Item> : RecyclerView.ViewHolder {

    private var itemViewBinder: ItemViewBinder<ViewType, ItemType>? = null

    constructor(root: ViewGroup,
                itemViewBinder: ItemViewBinder<ViewType, ItemType>) : super(itemViewBinder.getView(root.context)) {
        this.itemViewBinder = itemViewBinder
        itemView.asRow()
    }

    constructor(root: ViewGroup,
                emptyViewBinder: EmptyViewBinder) : super(emptyViewBinder.getView(root.context)) {
        this.itemViewBinder = null
        itemView.apply {
            setTag(EMPTY_TYPE, EMPTY_TYPE)
            this.fullSize()
        }
    }

    fun bind(item: ItemType, previousItem: ItemType?) {
        itemViewBinder?.bind(itemView as ViewType, item, previousItem)
    }
}
